### PR TITLE
L2 genesis - refactor to use solidity script

### PR DESCRIPTION
**Pull request by protolambda**
**Description**

Rebased version of https://github.com/ethereum-optimism/optimism/pull/9792

Taking ownership over this PR to add ProxyAdmin, Preinstalls, and integrate into op-e2e / devnet testing.

TODO:
- [x] Work from Mark to revive and improve changes from Wyatt to add missing predeploys to the script.
- [x] Insert `ProxyAdmin` into L2 predeploys
- [x] Support Preinstalls in solidity script
  - [x] `Permit2`
  - [x] `Create2Deployer`
  - [x] `DeterministicDeploymentProxy`
  - [x] `EntryPoint (4337)`
  - [x] `Safe_v130`
  - [x] `SenderCreator`
  - [x] `SafeSingletonFactory`
  - [x] `MultiSend_v130`
  - [x] `MultiCall3`
  - [x] `SafeL2_v130`
  - [x] `MultiSendCallOnly_v130`
- [x] Apply ecotone
- [x] Update `Setup` solidity contract to use this, rather than Go FFI
- [ ] ~~Re-use superchain-registry code path to build full `genesis.json` and `rollup.json``~~ Minimized op-chain-ops a lot already, reusing registry requires more diff, probably better for a different PR.
- [x] Integrate into op-e2e
- [x] Integrate into devnet (running in CI, verifying results still)

**Metadata**

Fix https://github.com/ethereum-optimism/protocol-quest/issues/151

### CODE REVIEW

File: packages/contracts-bedrock/scripts/L2Genesis.s.sol
Comment (tynes) [inline packages/contracts-bedrock/scripts/L2Genesis.s.sol]: You can represent this as `bytes32(0)` more cleanly

Comment (protolambda) [reply packages/contracts-bedrock/scripts/L2Genesis.s.sol]: I thought about it, but this is consistent with all the other slot declarations. I'll change it to `bytes32(0)` if you like that better though.

Comment (tynes) [reply packages/contracts-bedrock/scripts/L2Genesis.s.sol]: I think it would be much cleaner to do `bytes32(0)`, less characters, easier to update + verify correctness, the old stuff was just there and should be updated to new syntax if necessary as well

File: op-node/cmd/genesis/cmd.go
Comment (tynes) [inline op-node/cmd/genesis/cmd.go]: This should be required?

File: packages/contracts-bedrock/src/libraries/Predeploys.sol
Comment (tynes) [reply packages/contracts-bedrock/src/libraries/Predeploys.sol]: needs a prefix of the contract name to fix this like `Predeploys: unnamed predeploy`

File: packages/contracts-bedrock/scripts/L2Genesis.s.sol
Comment (tynes) [inline packages/contracts-bedrock/scripts/L2Genesis.s.sol]: you can technically delete this `setUp` as it does nothing in addition to the `super.setUp`

**Comment by protolambda** [comment]
rebased on develop, fixed an ugly few solidity test integration issues, devnet should work (will watch for CI) and going to work on op-e2e now

File: packages/contracts-bedrock/test/L2Genesis.t.sol
Comment (tynes) [inline packages/contracts-bedrock/test/L2Genesis.t.sol]: We should delete the extra unused functions in this file

File: packages/contracts-bedrock/test/L2Genesis.t.sol
Comment (tynes) [inline packages/contracts-bedrock/test/L2Genesis.t.sol]: dead comment

File: packages/contracts-bedrock/test/L2Genesis.t.sol
Comment (protolambda) [reply packages/contracts-bedrock/test/L2Genesis.t.sol]: Maybe we should add some addiitonal checks using these functions instead? What ideas did you have for testing in the original draft?

File: op-chain-ops/genesis/layer_two_test.go
Comment (tynes) [inline op-chain-ops/genesis/layer_two_test.go]: We can delete this test in the future, the solidity tests should cover this

File: op-chain-ops/genesis/layer_two_test.go
Comment (tynes) [inline op-chain-ops/genesis/layer_two_test.go]: Its good to leave this test for now but we should def move away from it in the future and rely completely on the solidity test coverage which i do believe almost already covers this besides the "use gov token" flag

File: packages/contracts-bedrock/scripts/Config.sol
Comment (tynes) [inline packages/contracts-bedrock/scripts/Config.sol:95]: In the future we should move away from using this function but fine to keep for now as it would be a breaking change, its just not easy to maintain and also results in implicit behavior that is non standard based on chain id as not all chain ids are considered "known"

File: packages/contracts-bedrock/scripts/L2Genesis.s.sol
Comment (tynes) [inline packages/contracts-bedrock/scripts/L2Genesis.s.sol:127]: This creates a really good way for us to standardize on the initial release of bedrock inside of `setPredeployImplementations`, we can copy/paste the initial release bytecode to be used there. This would mean we want a "development" output mode which uses the contracts based on the source code when the command runs. No need to do this right now but curious what you think of this approach?

File: packages/contracts-bedrock/scripts/L2Genesis.s.sol
Comment (tynes) [inline packages/contracts-bedrock/scripts/L2Genesis.s.sol:280]: it would probably be good to turn the reset into an internal function and reuse its impl everywhere. You can see how forge-std does it [here](https://github.com/foundry-rs/forge-std/blob/5dd1c68131ddd3c89ef169666eb262b92e90507c/src/StdCheats.sol#L547)

we don't need to go as far as they do, for us setting the balance to 0 technically isn't required as we don't have balance in this context

File: packages/contracts-bedrock/scripts/L2Genesis.s.sol
Comment (tynes) [inline packages/contracts-bedrock/scripts/L2Genesis.s.sol]: small semgrep fixes

File: packages/contracts-bedrock/scripts/L2Genesis.s.sol
Comment (tynes) [inline packages/contracts-bedrock/scripts/L2Genesis.s.sol]: You can do `address.code.length`, no need for assembly

File: packages/contracts-bedrock/scripts/L2Genesis.s.sol
Comment (tynes) [inline packages/contracts-bedrock/scripts/L2Genesis.s.sol]: semgrep wants the contract name prefixed like so `L2Genesis: ...`

File: packages/contracts-bedrock/scripts/L2Genesis.s.sol
Comment (tynes) [inline packages/contracts-bedrock/scripts/L2Genesis.s.sol:474]: This technically doesn't fully activate ecotone, we would be able to make good use of https://github.com/foundry-rs/foundry/pull/4931 to accomplish this easily. There are a few paths forward here:

- get the cheatcode into forge and just hardcode the rlp encoded tx. This would also require building foundry with optimism feature flag since its a deposit tx
- defining a solidity version of the "message" that is represented by the upgrade deposit tx

I lean towards just defining the messages given we would need to use a special foundry with optimism build flag and its unclear if that would even work e2e. Not ideal that its not 1:1 with the go, we would prob want a differential test long term

File: packages/contracts-bedrock/test/L2Genesis.t.sol
Comment (tynes) [reply packages/contracts-bedrock/test/L2Genesis.t.sol]: The problem was that they were prohibitively slow

Comment (tynes) [reply packages/contracts-bedrock/test/L2Genesis.t.sol]: It will take forever to run if we do something like "iterate over each account and get its balance" so we should say "jq get all accounts that have non zero balance" and "jq get all accounts that have 1 wei" and assert that the number is the same, this will say there are definitely only that number of accounts with 1 wei and it should be the expected number which iirc is 256

File: packages/contracts-bedrock/test/Preinstalls.t.sol
Comment (tynes) [inline packages/contracts-bedrock/test/Preinstalls.t.sol:27]: Can't we update it? hmm

File: packages/contracts-bedrock/test/Preinstalls.t.sol
Comment (tynes) [inline packages/contracts-bedrock/test/Preinstalls.t.sol]: We could set the deploy type to latest to fix this?

File: packages/contracts-bedrock/scripts/L2Genesis.s.sol
Comment (tynes) [reply packages/contracts-bedrock/scripts/L2Genesis.s.sol:127]: Ah this dev mode is local delta

File: packages/contracts-bedrock/scripts/L2Genesis.s.sol
Comment (tynes) [inline packages/contracts-bedrock/scripts/L2Genesis.s.sol:139]: How would we express just ecotone? Should we make mode more of a chainspec thing where its the hardfork name?

File: packages/contracts-bedrock/scripts/L2Genesis.s.sol
Comment (tynes) [inline packages/contracts-bedrock/scripts/L2Genesis.s.sol]: smol nit: i don't think we need the `deployed` prefix, could just add a comment on the function name

File: packages/contracts-bedrock/src/libraries/Predeploys.sol
Comment (tynes) [inline packages/contracts-bedrock/src/libraries/Predeploys.sol:86]: I like this function and it can be used in multiple places

File: packages/contracts-bedrock/scripts/L2Genesis.s.sol
Comment (protolambda) [reply packages/contracts-bedrock/scripts/L2Genesis.s.sol:139]: ecotone is the default, latest fork, without any suffix.

File: packages/contracts-bedrock/scripts/L2Genesis.s.sol
Comment (protolambda) [reply packages/contracts-bedrock/scripts/L2Genesis.s.sol:474]: The upgrade txs don't apply, as it deploys the same contracts that are already in place by default. The 4788 contract is installed as preinstall, independent of hardfork. And then only the `setL1BlockValuesEcotone` call remains to do here.

File: packages/contracts-bedrock/test/Preinstalls.t.sol
Comment (protolambda) [reply packages/contracts-bedrock/test/Preinstalls.t.sol:27]: Bug in the upstream library test tooling that has existed forever. Don't think we need to change it. The mainnet L1 bytecode serves as a template, that passes tests when using chainID=1 itself, well enough.

File: packages/contracts-bedrock/scripts/L2Genesis.s.sol
Comment (protolambda) [reply packages/contracts-bedrock/scripts/L2Genesis.s.sol:280]: This is not a full reset, but rather the reset of just the specific contract that was temporarily deployed for genesis purposes. Open to suggestions for how the reset should be, but don't think there's a lot to reduce.

**Comment by tynes** [comment]
Perhaps we want to improve the log `retrieving L1 deployments from artifacts` so it includes the path that its reading from? Looks like CI is failing on this right now

File: packages/contracts-bedrock/src/libraries/Predeploys.sol
Comment (tynes) [inline packages/contracts-bedrock/src/libraries/Predeploys.sol:111]: Is that valid solidity?

**Comment by protolambda** [comment]
@tynes there are already sufficient logs / forge debug data to see the exact issue:
```
    ├─ [0] VM::getDeployedCode("Proxy.sol:Proxy") [staticcall]
    │   └─ ← [Revert] No matching artifact found
    └─ ← [Revert] No matching artifact found
```
It fails to retrieve the `Proxy` artifact.

File: packages/contracts-bedrock/scripts/L2Genesis.s.sol
Comment (protolambda) [reply packages/contracts-bedrock/scripts/L2Genesis.s.sol:127]: bad suggestion

**Comment by protolambda** [comment]
I think the `Proxy.sol` not found error, even though it shows up in the artifacts directory, is related to: https://github.com/foundry-rs/foundry/issues/7569

File: op-chain-ops/genesis/layer_two.go
Comment (trianglesphere) [inline op-chain-ops/genesis/layer_two.go:42]: How confident are we that the chainID will always be at this bytecode location?

Comment (protolambda) [reply op-chain-ops/genesis/layer_two.go:42]: 100%, the bytecode of permit2 is hardcoded, not compiled from source. And permit2 itself is not changing anymore either. Solidity unit tests ensure that the behavior is correct with different chain IDs at this offset.

**Review by 0xfuturistic — APPROVED** [review]
lgtm after resolving conflict

**Comment by protolambda** [comment]
Rebased + fixed devnet python merge conflict + added gas snapshot + added Proxy import into L2Genesis script to try and fix missing artifact flake / forge issue.

**Comment by protolambda** [comment]
Blocked by https://github.com/foundry-rs/foundry/issues/7732

**Comment by protolambda** [comment]
Rebased and squashed my commits, `forge script` fix should now work in CI. Really looking forward to get this refactor merged. Will be ooo this week otherwise.

**Comment by tynes** [comment]
Accidentally closed this with my commit message

**Review by tynes — APPROVED** [review]

**Pull request merged**
